docs/compactor.md

```diff
@@ -0,0 +1,36 @@
+# Compaction for ETCD
+
+## Current Problem
+To ensure recoverability of ETCD, backups of the database are taken at regular interval.
+Backups are of two types: Full Snapshots and Incremental Snapshots. 
+
+**Full Snapshots:** Full snapshots are memory snapshots of the database. ETCD is an in-memory database, so the size of the database keeps growing with time. As a result, the size of the memory snapshots grows with time. Full Snapshots are taken after some large intervals.
+
+**Incremental Snapshots:** Incremental Snapshots are collection of commands on ETCD database, obtained through running WATCH API Call on ETCD. After some short intervals, all the commands that are accumulated through WATCH API Call are saved in a file and named as Incremental Snapshots.
+
+### **Recovery from the Snapshots:**
+**Recovery from Full Snapshots:** As the full snapshots are just the snapshot of memory, it takes less time to recover from a full snapshot. Recovery only needs to copy the provided full snapshot file from disk to memory. ETCD provides API Call to restore the database from a full snapshot file.
+
+**Recovery from Incremental Snapshots:** Delta snapshots are collection of retrospective ETCD commands. So, to restore from Incremental snapshot file, the commands from the file are needed to be applied sequentially on ETCD database through ETCD Put/Delete API calls. As it is heavily dependent on ETCD calls sequentially, restoring from Incremental Snapshot files can take long if there are numerous commands captured in Incremental Snapshot files.
+
+Delta snapshots are applied on top of running ETCD database. So, if there is inconsistency between the state of database at the point of applying and the state of the database when the delta snapshot commands were captured, restoration will fail.
+
+Currently, in Gardener setup, ETCD is restored from the last full snapshot and then the delta snapshots, which were captured after the last full snapshot.
+
+Now, the problem is that as times go by the size of the ETCD database increases. And the problem becomes more troubling as in the case of ETCD every change in ETCD database creates a new revision of the database. So, the Full Snapshot size also increases. But without compaction, the older revisions of ETCD database are also included in the snapshots. So even during restoration, ETCD is restored with older uncompacted huge database. Eventually, ETCD gets Out of Memory.
```

Comment:
```suggestion

The main problem with this is that the complete restoration time can be unacceptably long if the rate of change coming into the etcd database is quite high because there are large amount delta snapshots to be applied sequentially.

A secondary problem is that, though auto-compaction is enabled for etcd, the auto-compaction period it is not quick enough to compact all the changes from the incremental snapshots being re-applied during the relatively short period of time of restoration (as compared to the actual period of time when the incremental snapshots were accumulated). This may lead to the etcd pod (the backup-restore sidecar container, to be precise) to run out of memory and/or storage space even if it is sufficient for normal operations.
```